File: measurementblock.m
classdef (Abstract) measurementblock < handle
    %MEASUREMENTBLOCK A generic Kalman Filter Measurement Block class
    % This class is an abstract implementation of the measurement block of 
    % a Kalman Filter. This class is used to form the observation,
    % innovation covariance, and measurement covariance matrices. It also
    % stands to correct an estimate with incoming measurements using the
    % measurement model.
    %
    % Author: Walter Livingston
    
    properties
        meas_sigmas     (1,:) double = 0;       % Vector of State Sigmas
        H               (:,:) double = NaN;     % Observation Matrix
        aux             (1,1) struct            % Struct of Aux Data
    end
    
    methods
        function obj = measurementblock(meas_sigmas)
            obj.meas_sigmas = meas_sigmas;
        end
    end

    methods(Abstract)
        [H]         = updateObservationMatrix(obj, x_, y_);
        [R]         = calcMeasurementCovarianceMatrix(obj);
        [S]         = calcInnovationCovarianceMatrix(obj, x_, P_);
        [yhat, H]   = update(obj, x_, y_, relinearize);
    end

    methods
        function [obj] = processAuxData(obj, aux)
            obj.aux = aux;       
        end
        function [x, X] = applyError(obj, x_, X_)
            error(['process_aux_data() for the state block has not ' ...
                'been implemented!']);        
        end
    end
end

File: modular_kf/measurement_block.py
import numpy as np
from abc import ABC, abstractmethod
from modular_kf.helpers import *

class MeasurementBlock(ABC):
    meas_sigmas: vec = vec0
    H: mat = mat0
    aux: Aux = Aux(x = vec0, X = vec0, y = vec0)

    def __init__(self, meas_sigmas):
        self.meas_sigmas = meas_sigmas

    @abstractmethod
    def updateObservationMatrix(self, x_: vec, y_:vec) -> mat:
        pass

    @abstractmethod
    def calcMeasurementCovarianceMatrix(self) -> mat:
        pass

    @abstractmethod
    def calcInnovationCovarianceMatrix(self) -> mat:
        pass

    @abstractmethod
    def update(self, x_: vec, y_: vec, relinearize: bool = True) -> tuple[vec, mat]:
        pass

    @abstractmethod
    def applyError(self, x_: vec, X_: vec) -> estWithNominal:
        pass

    def processAuxData(self, aux_: Aux):
        self.aux = aux_